Make ActiveSupport frozen-string-literal friendly.

pat opened the pull request:
As I'm sure many folks are aware, MRI 2.3 added the ability to have [all string literals frozen by default](https://wyeworks.com/blog/2015/12/1/immutable-strings-in-ruby-2-dot-3), which has performance/memory usage benefits. It will become the default behaviour in MRI 3.0, though that is quite some time away!

This PR covers what's needed to make ActiveSupport play nicely when this feature is enabled. The test suite as it currently stands will continue to pass, there are no behaviour changes. Yes, it's for ActiveSupport only, but I figure it's a good place to start.

If this change is appreciated, then I can work further on the rest of the Rails gems to ensure they're working with frozen string literals as well. And if you want to have Travis run the tests with checks for this behaviour, then using pragmater and a test script can help with that (such as what I've done in [my own gems](https://github.com/pat/thinking-sphinx/blob/develop/bin/literals) and suggested [for others](https://github.com/plataformatec/devise/pull/4566)).

## Running the tests with frozen string literals enabled

Of course, if you want to run the test suite with frozen string literals enabled, you'll need the following locally:
* Use the latest commits from dalli, rather than the latest gem release.
* Use my PR for Builder, rather than the latest gem release: https://github.com/tenderlove/builder/pull/6
* Use Nokogiri 1.8.0

Beyond that, all external dependencies (at least, to the extent they’re used by ActiveSupport) play happily with frozen string literals enabled by default in MRI 2.4 (which can be done so by the `RUBYOPT` environment variable: `RUBYOPT="--enable-frozen-string-literal"`).

pat committed:
Make ActiveSupport frozen string literal friendly.

The ActiveSupport test suite only passes currently if it uses the latest unreleased commits for dalli, and a patch for Builder:
https://github.com/tenderlove/builder/pull/6

Beyond that, all external dependencies (at least, to the extent they’re used by ActiveSupport) are happy, including Nokogiri as of 1.8.0.

tenderlove commented:
I think it's good if we make Rails internals "frozen string" friendly.  Though I'd prefer if we used `.dup` instead of `String.new`.  I feel like I'm having Java flashbacks.  😉

pat commented:
Fair enough :) I meant to comment on the different approaches (`String.new` vs `dup` vs `StringIO` vs `+=` in place of `<<`). I'll switch my changes to `dup`.

pat committed:
Make ActiveModel frozen string literal friendly.

Includes two external changes because they're referenced within the ActiveModel test suite.

koic commented:
Incidentally, the encoding of the empty string generated by `String.new` without arguments is not UTF-8 but ASCII-8BIT.

```ruby
String.new.encoding     # => #<Encoding:ASCII-8BIT>
String.new("").encoding # => #<Encoding:UTF-8>
"".dup.encoding         # => #<Encoding:UTF-8>
"".encoding             # => #<Encoding:UTF-8>
```

I also think that it is preferable to use the `String#dup` rather than `String.new` so that this difference does not cause unexpected problems.
(There is no problem because `String.new` without arguments did not exist in this PR.)

pat committed:
Make ActiveRecord frozen string literal friendly.

pat committed:
Make ActionMailer frozen string literal friendly.

pat committed:
Make ActionView frozen string literal friendly.

Plus a couple of related ActionPack patches.

pat commented:
Thanks for that note @koic, good to know :)

I've gone ahead and made changes to ActiveModel, ActiveRecord, ActionMailer and ActionView. ActionJob didn't need any changes. For ActionCable, ActionPack and Railties, there are underlying dependencies that need patching (redis, rack, and thor, respectively), so I'll look into those tomorrow.

If anyone with particular focus/expertise on the libraries I _have_ patched would like to review, please go ahead!

pat commented:
Also, @tenderlove: I updated my Builder patch to use `dup` as well 👍

rafaelfranca commented:
Can we add the magic comment to the files that were fixed?

pat commented:
@rafaelfranca we could, but that's not going to help stop bad uses of frozen string literals appear in all the other files. We could add the comment to _all_ current files, but still, any new files that get added to Rails would need to have it as well - which may be acceptable?

Using either [pragmater](https://github.com/plataformatec/devise/pull/4566/files) or [`RUBYOPT`](https://github.com/pat/sliver/blob/master/.travis.yml#L3) helps avoid regressions more reliably, I feel. Certainly, if all dependencies are frozen-string-literal friendly, then `RUBYOPT` is the simpler approach, but I'm a little wary of that given the number of dependencies Rails (and its test suite) relies upon.

matthewd commented:
@pat thanks for this!

I think the comment is important for these files in particular because otherwise `"".dup` will allocate two strings when the global option isn't set.  Even so, we should consider whether any of these are *particularly* hot -- the magic comment can't help us for 2.2.

@rafaelfranca perhaps we should consider just adding the comment everywhere, and enabling the relevant rubocop rule. Once we (read: @pat :wink:) have done the work to find all the strings that *need* a `dup`, the rest are all neutral at best, or otherwise a free [trivial] performance gain.

matthewd commented:
Bonus thought: even with the magic comment in every file, that wouldn't (AFAIK) cover eval calls, of which we have a few -- though most won't contain strings.

Probably narrow enough that we can ignore enforcement for now, until we can go full RUBYOPT on our test suite.

tenderlove commented:
> @rafaelfranca perhaps we should consider just adding the comment everywhere, and enabling the relevant rubocop rule. Once we (read: @pat 😉) have done the work to find all the strings that need a dup, the rest are all neutral at best, or otherwise a free [trivial] performance gain.

+1.  On that note, are there actual benchmarks that show the frozen string default helping (especially for a real app)?  Last thing I saw was a modest speed increase for micro benchmarks.  But the last tests I saw were probably 2 years ago when we were discussing the feature on ruby-core.

I still think we should move forward with frozen string literals (because I like them), I'm just genuinely curious about real performance.

rafaelfranca commented:
> perhaps we should consider just adding the comment everywhere, and enabling the relevant rubocop rule

That is the idea. I did that already in arel/
